activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -55,6 +55,13 @@ def build_bind_attribute(column_name, value)
       Arel::Nodes::BindParam.new(attr)
     end
 
+    def build_query_attribute(column_name, value)
+      bind = build_bind_attribute(column_name, value)
+      bind.nil?
```

Comment:
Why do we have the random `nil?` check?

Reply:
This intend to call `QueryAttribute#value_for_database` in `QueryAttribute#nil?` in `BindParam` to check `Type::Integer#ensure_in_range` before building arel predicate.

https://github.com/rails/rails/blob/51c5bcbb39777b3861a44042a598bfb07d4e1fc1/activerecord/lib/active_record/relation/query_attribute.rb#L20-L23

Reply:
See my second comment. I'd like to see this made more explicit in the code. This is too spooky-action-at-a-distance for my taste.

Reply:
Especially since `nil?` there is basically just a hack to satisfy Arel that I'd like to remove at some point.

Reply:
hm... how about here?

```diff
diff --git a/activerecord/lib/active_record/relation/predicate_builder.rb b/activerecord/lib/active_record/relation/predicate_builder.rb
index c69a238eff..56e632f0d7 100644
--- a/activerecord/lib/active_record/relation/predicate_builder.rb
+++ b/activerecord/lib/active_record/relation/predicate_builder.rb
@@ -58,7 +58,13 @@ def build_bind_attribute(column_name, value)
 
     def build_query_attribute(column_name, value)
       bind = build_bind_attribute(column_name, value)
-      bind.nil?
+      attr = bind.value
+
+      # Checking whether an integer value is within the range
+      if attr.type.type == :integer && attr.value_before_type_cast
+        attr.value_for_database
+      end
+
       bind
     rescue ::RangeError
     end
```

Reply:
That's slightly more obvious about what it's doing, but still feels both over-specialised and too vague at the same time. My first thought was maybe we want a `bind.value.impossible?` or similar, but that isn't actually sufficient for our needs: see the new tests I proposed.

I'm not excited about this method name, either... particularly the fact we're losing "bind", which seems important to understand while reading callers. But that's better revisited after the changes for the new tests too.